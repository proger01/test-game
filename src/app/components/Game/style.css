.mainText {
  color: white;
  background-color: #be444c;
  border-radius: 4px;
  font-weight: bold;
  font-size: 42px;
  margin: 12px;
}

.subMainText {
  font-size: 24px;
  background-color: #be444c;
  border-radius: 4px;
  color: white;
  margin: 12px;
}

.variantsBlock {
  margin: 24px 0;
  display: flex;
  align-items: center;
  justify-content: center;
  font-size: 18px;
}

.variant {
  cursor: pointer;
  border: solid 1px black;
  border-radius: 6px;
  width: 300px;
  height: 72px;
  color: white;
  background-color: #be444c;
  display: flex;
  align-items: center;
  justify-content: center;
  position: relative;
}

.variant:hover {
  background-color: #a73c43;
  color: yellow;
}

.variant:active {
  outline: none;
}

.variantAnswer {
  max-width: 80%;
}

.fontawesome {
  width: 10%;
  position: absolute;
  right: 6px;
  float: right;
}

.image {
  max-width: 100%;
  width: 820px;
  height: 460px;
  object-fit: cover;
  border-radius: 8px;
  margin: 20px 0;
}

.question {
  font-size: 22px;
  color: white;
  background-color: #be444c;
}

.button {
  background-color: #28a745;
  width: 100px;
  height: 48px;
  color: white;
  border: none;
  border-radius: 8px;
}

.button:hover {
  background-color: #228a3a;
}

.button:active {
  outline: none;
}

.result {
  width: 280px;
  background-color: #be444c;
  color: white;
  margin: 6px auto;
}

.insteadResult {
  height: 90px;
}

.endResult {
  color: white;
  background-color: #be444c;
  width: 280px;
  margin: 6px auto;
  font-size: 30px;
}

/* * * * * * **/
/* Responsive */
/* * * * * * **/

@media (max-width: 1024px) {
  .variantsBlock {
    display: block;
    flex-direction: column;
  }

  .variant {
    display: flex;
    justify-content: flex-start;
    width: 100%;
    height: 50px;
  }

  .variantAnswer {
    margin-left: 20px;
  }
}

@media (max-width: 767px) {
  .mainText {
    font-size: 32px;
  }

  .subMainText {
    font-size: 20px;
  }
}

@media (max-width: 480px) {
  .mainText {
    font-size: 28px;
  }

  .subMainText {
    font-size: 16px;
  }

  .question {
    font-size: 18px;
  }
}
