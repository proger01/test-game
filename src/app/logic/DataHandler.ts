import { TestData } from '../models/TestData';

export class DataHandler {
  public data: TestData;

  constructor(data: TestData) {
    this.data = data;
  }

  public getTitle = (): string => {
    return this.data.title;
  };

  public getSubTitle = (): string => {
    return this.data.subTitle;
  };

  public getFirstImage = (): string => {
    return this.data.firstImage;
  };

  public getLastImage = (): string => {
    return this.data.lastImage;
  };

  public getQuestionImage = (questionNumber: number): string => {
    return this.data.questions[questionNumber].image;
  };

  public getQuestionTitle = (questionNumber: number): string => {
    return this.data.questions[questionNumber].title;
  };

  public getQuestionAnswers = (questionNumber: number): string[] => {
    return this.data.questions[questionNumber].answers;
  };

  public getQuestionCorrectAnswer = (questionNumber: number): string => {
    return this.data.questions[questionNumber].correctAnswer;
  };
}
