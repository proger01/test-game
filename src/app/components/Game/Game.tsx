import * as React from 'react';
import { DataHandler } from '../../logic/DataHandler';
import { faCheckCircle, faTimesCircle } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import styles from './style.css';

enum TestsQuantity {
  Start = -1,
  Quantity = 10,
}

interface GameProps {
  testData: DataHandler;
}

interface GameState {
  correctAnswers: number;
  isCorrectAnswer: boolean;
  showAnswer: boolean;
  testNumber: number;
}

export class Game extends React.Component<GameProps, GameState> {
  constructor(props: GameProps) {
    super(props);

    this.state = {
      correctAnswers: 0,
      isCorrectAnswer: false,
      showAnswer: false,
      testNumber: -1,
    };
  }

  public onSubmit = (): void => {
    this.setState({
      isCorrectAnswer: false,
      showAnswer: false,
      testNumber: this.state.testNumber + 1,
    });
  };

  public onSelect = (e: any): void => {
    if (
      e.target.textContent === this.props.testData!.getQuestionCorrectAnswer(this.state.testNumber)
    ) {
      this.setState({ isCorrectAnswer: true, correctAnswers: this.state.correctAnswers + 1 });
    }
    this.setState({ showAnswer: true });
  };

  public getVariantAnswers = () => {
    const answers = this.props.testData!.getQuestionAnswers(this.state.testNumber);
    if (answers.length) {
      return answers.map((answer: string, i: number) => {
        return (
          <button
            className={styles.variant}
            key={i}
            onClick={this.onSelect}
            disabled={this.state.showAnswer}
          >
            <div className={styles.variantAnswer}>{answer}</div>
            {!this.state.showAnswer ? null : (
              <div className={styles.fontawesome}>
                <FontAwesomeIcon
                  icon={
                    answer === this.props.testData!.getQuestionCorrectAnswer(this.state.testNumber)
                      ? faCheckCircle
                      : faTimesCircle
                  }
                />
              </div>
            )}
          </button>
        );
      });
    }
    return null;
  };

  public render() {
    return (
      <div>
        <h2 data-testid="title" className={styles.mainText}>
          {this.props.testData!.getTitle()}
        </h2>
        <h5 data-testid="sub-title" className={styles.subMainText}>
          {this.props.testData!.getSubTitle()}
        </h5>
        {this.mainRender()}
      </div>
    );
  }

  public mainRender = () => {
    if (this.state.testNumber === TestsQuantity.Start) {
      return this.renderStartGame();
    } else if (this.state.testNumber < TestsQuantity.Quantity) {
      return this.renderGame();
    }
    return this.renderEndGame();
  };

  public renderStartGame = () => {
    return (
      <div>
        <img className={styles.image} src={this.props.testData!.getFirstImage()} alt="img" />
        <div>
          <button data-testid="start-button" className={styles.button} onClick={this.onSubmit}>
            Начать
          </button>
        </div>
      </div>
    );
  };

  public renderGame = () => {
    return (
      <div>
        <img
          className={styles.image}
          src={this.props.testData!.getQuestionImage(this.state.testNumber)}
          alt="img"
        />
        <section>
          <h5>
            <strong data-testid="question" className={styles.question}>
              {this.props.testData!.getQuestionTitle(this.state.testNumber)}
            </strong>
          </h5>
          <div data-testid="variants" className={styles.variantsBlock}>
            {this.getVariantAnswers()}
          </div>
        </section>
        <div>
          {this.state.showAnswer ? (
            this.afterAnswer()
          ) : (
            <div className={styles.insteadResult}></div>
          )}
        </div>
      </div>
    );
  };

  public renderEndGame = () => {
    const correctAnswers = this.state.correctAnswers;
    const result = correctAnswers < 5 ? 'Можно лучше' : correctAnswers < 8 ? 'Сойдет' : 'Красава!';

    return (
      <div>
        <img className={styles.image} src={this.props.testData!.getLastImage()} alt="img" />
        <div className={styles.endResult}>
          <div>{`${correctAnswers}/10`}</div>
          <p>{result}</p>
        </div>
      </div>
    );
  };

  public afterAnswer = () => {
    const result = this.state.isCorrectAnswer
      ? 'Твой вариант правильный!'
      : 'Твой вариант неправильный!';
    return (
      <div>
        <div data-testid="result" className={styles.result}>
          {result}
        </div>
        <button data-testid="continue-button" className={styles.button} onClick={this.onSubmit}>
          Продолжить
        </button>
      </div>
    );
  };
}
