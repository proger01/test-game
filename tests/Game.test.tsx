import * as React from 'react';
import { testData } from './testData';
import { DataHandler } from '../src/app/logic/DataHandler';
import { render, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Game } from '../src/app/components/Game/Game';
import renderer from 'react-test-renderer';

const data: DataHandler = new DataHandler(testData);

const pressStartButton = () => {
  render(<Game testData={data} />);
  const startButton = screen.getByTestId('start-button');
  userEvent.click(startButton);
};

describe('Start page', () => {
  it('renders without crashing', () => {
    render(<Game testData={data} />);
  });

  it('renders the correct content of the start page', () => {
    render(<Game testData={data} />);
    const title = screen.getByTestId('title');
    const subTitle = screen.getByTestId('sub-title');
    const startButton = screen.getByTestId('start-button');

    expect(title.textContent).toBe(data.getTitle());
    expect(subTitle.textContent).toBe(data.getSubTitle());
    expect(startButton.textContent).toBe('Начать');
  });

  it("click the button 'Начать'", () => {
    render(<Game testData={data} />);
    const startButton = screen.getByTestId('start-button');
    fireEvent.click(startButton);
  });

  it('matches snapshot', () => {
    const component = renderer.create(<Game testData={data} />);
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
});

describe('First question', () => {
  const getVariantsArray = (htmlElement: HTMLElement): any => {
    return Array.from(htmlElement.childNodes).map((item) => item.textContent);
  };

  it("without crashing after clicking 'Начать' button", () => {
    pressStartButton();
  });

  it('renders the correct content of the 1st question page before selection a variant', () => {
    pressStartButton();

    const title = screen.getByTestId('title');
    const subTitle = screen.getByTestId('sub-title');
    const questionTitle = screen.getByTestId('question');
    const variantsArray = getVariantsArray(screen.getByTestId('variants'));

    expect(title.textContent).toBe(data.getTitle());
    expect(subTitle.textContent).toBe(data.getSubTitle());
    expect(questionTitle.textContent).toBe(data.getQuestionTitle(0));
    for (let i = 0; i < variantsArray.length; i++) {
      expect(variantsArray[i]).toBe(data.getQuestionAnswers(0)[i]);
    }
  });
  it('renders the correct answer', () => {
    pressStartButton();

    const variantsArray = getVariantsArray(screen.getByTestId('variants'));
    userEvent.click(variantsArray[2]);

    const result = screen.getByTestId('result');
    expect(result.textContent).toBe('Твой вариант правильный!');
  });
});
