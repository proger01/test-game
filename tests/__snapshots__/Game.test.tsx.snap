// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Start page matches snapshot 1`] = `
<div>
  <h2
    className="mainText"
    data-testid="title"
  >
    Тест: насколько хорошо ты знаешь города Украины
  </h2>
  <h5
    className="subMainText"
    data-testid="sub-title"
  >
    Эй, знаток родных краев, готов(-а) проверить насколько хорошо ты разбираешься в вопросе? Мы подготовили тест, который поможет это узнать. Не забудь написать свой результат в комментариях!
  </h5>
  <div>
    <img
      alt="img"
      className="image"
      src="//tripmydream.cc/travelhub/travel/block_gallery/10/8067/default_108067.jpg"
    />
    <div>
      <button
        className="button"
        data-testid="start-button"
        onClick={[Function]}
      >
        Начать
      </button>
    </div>
  </div>
</div>
`;
